---
title: "Stratified analysis Franâ€™s dataset"
author: "Lan Ngo"
date: "7/14/2017"
output: 
  html_document:
    toc: true
    theme: journal
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, tidy = TRUE, tidy.opts=list(width.cutoff=50), eval=TRUE)
```

```{r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Transcriptomic signatures of sepsis severity

## Abstract
Sepsis is a clinical syndrome defined by physiologic changes indicative of systemic inflammation, which are likely attributable to documented or suspected infection.
spectrum of disorders that result from infection by bacteria, viruses, fungi, or parasites or the toxic products of these microorganisms. 


Load all necessary libraries
```{r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
library(oligo)
library(limma)
library(annotate)
library(arrayQualityMetrics)
library(viridis)
library(pd.hugene.2.0.st)
library(hugene20sttranscriptcluster.db)
library(sva)
library(ggplot2)
library(gmodels)
library(knitr)
library(pander)
library(randomForest)
```


Loading saved results from previous run

``` {r eval = TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
setwd("/project/bhimeslab/sepsis")
load("run0720-2.RData")
```

## EDA for Aim 1

```{r eval = TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
library(ggplot2)
vars <- c("aim2_short", "race_ethnicity", "sex")
par(mfrow=c(1,3))
unlist(lapply(vars, function(var){
  tab <- data.frame(var=pData(sepsis.rma)[[var]],aim1=pData(sepsis.rma)$aim1_status)
  ggplot(data=tab, mapping = aes(fill=as.factor(var), x=as.factor(aim1)))+geom_bar(position = "fill")+labs(x="Aim 1", y = var, title=paste("Aim 1 status by", var))+coord_flip()
}))
```

Remove healthy controls and collapse pressor and fluid_responsive samples together
```{r eval = TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
sepsis.rma <- sepsis.rma[ ,-which(pData(sepsis.rma)$aim1_status=="control")]
pData(sepsis.rma)$aim1_status <- unlist(lapply(pData(sepsis.rma)$aim1_status, function(x){
  if (x=="no_shock"){x}
  else {"pressor_fluid"}
}))
```

Choose the aim, and then the covariants to be adjusted. Construct a dictionary of gene names for faster access in later runs.

```{r eval = TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
choices <- list(c("aim1_status", "race_ethnicity", "sex"), 
                c("aim1_status", "aim2_short"),
                c("aim2_short", "race_ethnicity", "sex"),
                c("aim1_status"))
groups_aim1 <- list(c("pressor_fluid", "no_shock"))
groups_aim2 <- list(c("viral", "bacterial"))

# Saved gene names for faster look-up later 
# gene_names <- sapply(row.names(sepsis.rma), function(x) 
#   mget(x, env=hugene20sttranscriptclusterSYMBOL, ifnotfound=NA)[[1]])
# names(gene_names) <- row.names(sepsis.rma)
```

Constructing full dataset with sepsis.rma, clinical variables 1:17
```{r eval = TRUE, echo = TRUE, results='asis',  message=FALSE, warning=FALSE}
omit <- c("study_id", "cel_file", "aim1_combined", "aim2_status", "race", "ethnicity", 
          "rna_extraction_batch", "Scan_Date")
names <- names(pData(sepsis.rma))[-which(names(pData(sepsis.rma)) %in% omit)]
full.data <- cbind(pData(sepsis.rma)[,-which(names(pData(sepsis.rma)) %in% omit)],
                   t(exprs(sepsis.rma)))
full.data.complete <- full.data.complete <- full.data[complete.cases(full.data),]
full.data.imputed <- full.data
full.data.imputed[,1:17] <- lapply(full.data.imputed[,1:17], function(x) as.numeric(as.factor(x)))
full.data.imputed[,1:17] <- rfImpute(aim1_status ~., data=full.data.imputed[,1:17])
```


We experimented with 2 unsupervised learning methods: PCA and k-means clustering to 
## Principle Component Analysis

# K-means clustering
Main function for getting top DE genes 

```{r eval = TRUE, echo = TRUE, results='asis',  message=FALSE, warning=FALSE}
analysis <- function(gset, choice, group, complete_cases){
  print(paste("Analysis for target:", choice[1], "adjusted for", choice[2:length(choice)]))
  if (complete_cases == TRUE){
    # Remove rows containing NA values
    throw <- function(cov){which(is.na(pData(gset)[[cov]])==TRUE)}
    rows_to_throw <- unlist(lapply(choice, throw))
    if (length(rows_to_throw)!=0){
        gset <- gset[,-rows_to_throw]
    }
  } else {
    # Create a new label "missing" to replace NA
    pData(gset)$aim2_short <- unlist(lapply(pData(gset)$aim2_short, function(x){
      if (is.na(x)==TRUE){"missing"}
      else {x}
    }))
  }
  
  # Cross-tabulation
  vars <- lapply(choice, function(var){pData(gset)[[var]]})
  joint <- table(unlist(vars))
  print(kable(t(as.matrix(joint)), caption=paste("Cross-tabulation for", choice[1]))) 
  
  # Making design matrix
  factors <- unlist(lapply(choice, function(x){
    paste0("factor(pData(gset)[[\"", x, "\"]])")}))
  design <- model.matrix(as.formula(paste("~ -1+", paste(factors, 
                                                         collapse = "+"))))
  colnames(design) <- lapply(colnames(design), function(x){
    marker <- gregexpr(pattern=")", x)[[1]][2]
    substr(x, (marker+1), nchar(x))
  })
  
  # Linear fit
  fit <- lmFit(gset, design)
  # Setting up contrast matrix for interested groups
  make_groups <- unlist(lapply(1:length(group), function(x) {
    paste0("group", x, "=", group[[x]][1],"-", group[[x]][2])}))
  sepsis.contrast <- eval(parse(text=paste("makeContrasts(", 
                                           paste(make_groups, collapse=","),
                                           ",levels = design)")))
  fit2 <- contrasts.fit(fit, sepsis.contrast)
  fit2 <- eBayes(fit2)
  
  # Getting top genes results based on groups 
  groups <- unlist(lapply(1:length(group), function(x){paste0("group",x)}))
  # A function to get the top genes result for each group
  get_results <- function(group){
    press_results <- topTable(fit2, coef = group , adjust="BH", num=Inf)
    N <- length(press_results$P.Value[press_results$P.Value <0.0001])
    if (N < 20){N <- 20}
    top.eset <- gset[row.names(exprs(gset)) %in%
                             row.names(gset)[1:N]]
    if (is.na(top.esets)==TRUE){top.esets <<- list(top.eset)}
    else {top.esets <<- list(top.esets, top.eset)}
    #Heatmap of top DE
    press_results$ID <- row.names(press_results)
    press_results$SYMBOL <- sapply(press_results$ID,
                                   function(x){gene_names[[x]]})
    colors <- rainbow(length(unique(top.eset[[choice[1]]])))
    status.colors <- unlist(lapply(as.numeric(factor(top.eset[[choice[1]]])),
                                   function(x){colors[x]}))
    
    # Store top results in a table
    print(kable(head(press_results, 15), 
                caption=paste("Top genes for", group)))

    heatmap(exprs(top.eset), col=viridis(256, option="B"), margins = c(5,10),
            ColSideColors=status.colors, labRow=press_results$SYMBOL, labCol=FALSE)
    par(lend = 1)           # square line ends for the color legend
    title(paste(choice[1], group, "adjusted for", paste(choice[-1], sep="and")))
    legend("topright",      # location of the legend on the heatmap plot
        legend = names(table(top.eset[[choice[1]]])), # category labels
        col = colors,  # color key
        lty= 1,             # line style
        lwd = 10            # line width
    )
  }
  lapply(groups, get_results)
}
```

## Aim 1 Analysis
Conduct differential gene expression analysis for Aim1, unadjusted model

```{r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
top.esets <- NA
analysis(sepsis.rma, choices[[4]], groups_aim1, complete_cases = TRUE)
top.eset <- top.esets[[1]]
```

Creating a data.frame instance of top.eset
```{r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
# Constructing the dataframe
clin.vars <- c("aim1_status", "aim2_short", "race_ethnicity", "sex")
top.data <- cbind(pData(top.eset)[names(pData(top.eset)) %in% clin.vars], t(exprs(top.eset)))
# Need to factorize all the clinical variables
lapply(clin.vars, function(var){
  top.data[[var]] <<- as.numeric(as.factor(top.data[[var]]))})

# Checking the dimension of top data
print(paste("Checking the dataset's dimension: ", 
          dim(top.data)[1], dim(top.data)[2], "\n"))
# Imputing missing values in data
library(randomForest)
top.data.imputed <- rfImpute(aim1_status ~ ., top.data)
names(top.data.imputed) <- make.names(names(top.data.imputed))
```
## Unsupervised methods
#### K-means clustering
First using k-means to see if there is any clusters 
```{r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
set.seed(1)
sepsis.cluster.2 <- kmeans(top.data.imputed[,-1], 2, nstart=20)
print("K-means clustering for aim1 with k=2")
print(table(sepsis.cluster.2$cluster,top.data.imputed$aim1_status))
sepsis.cluster.4 <- kmeans(top.data.imputed[,-1], 4, nstart=20)
print("K-means clustering for aim1 with k=4")
print(table(sepsis.cluster.4$cluster,top.data.imputed$aim1_status))
```

#### PCA
Use PCA to check batch effects with the scan date, then to see any useful clusters get formed
```{r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
sepsis.pca <- prcomp(top.data.imputed[,-1], center=TRUE, scale. = TRUE)
summary(sepsis.pca)
# Projecting data on the first 2 principal components
library(ggfortify)
print("Checking for batch effect with Scan Date")
top.data.scanDate <- top.data.imputed
top.data.scanDate$Scan_Date <- pData(top.eset)$Scan_Date 
autoplot(prcomp(top.data.scanDate[,-25],center = TRUE, scale. = TRUE), 
         data=top.data.scanDate, color=top.data.scanDate$Scan_Date, loadings = TRUE)
print("PCA to discover any relevant clusters for aim 1")
autoplot(prcomp(top.data.imputed[,-1], center = TRUE, scale. = TRUE),
         data = top.data.imputed, color = "aim1_status", loadings = TRUE)

```

## Linear Discriminant Analysis (LDA) based on unadjusted model
```{r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
library(MASS)
sepsis.lda <- lda(aim1_status ~., top.data.imputed)
sepsis.lda.values <- predict(sepsis.lda)
print(sepsis.lda)
```

## LDA histograms
1: Fluid_responsive 2: no_shock 3: pressor
``` {r eval=TRUE, message=FALSE, warning=FALSE}
print("Stacked histogram of LDA values for first PC")
ldahist(data=sepsis.lda.values$x[,1], g=sepsis.lda.values$class)

cat("Stacked histogram of LDA values for second PC")
ldahist(data=sepsis.lda.values$x[,2], g=sepsis.lda.values$class)

cat("Scatterplot of Discriminant functions")
plot(sepsis.lda.values$x[,1], sepsis.lda.values$x[,2])
text(sepsis.lda.values$x[,1], sepsis.lda.values$x[,2], sepsis.lda.values$class, pos=4, col='red')
```

We proceeded to conduct 3-fold CV with LDA for aim1 (pressor_fluid vs. no_shock)

```{r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
library(pROC)
library(caret)
lda_sepsis <- function(data, target, n){
  set.seed(1)
  folds <- createFolds(data[[target]], k = n, list = TRUE, returnTrain = FALSE)
  sens <- spec <- accu <- numeric(n)
  N <- nrow(data)
  preds <- obs <- vector(mode="numeric", length = N)
  offset <- 0
  for (i in 1:n){
    train <- data[-folds[[i]],]
    test <- data[folds[[i]],]
    train.lda <- eval(parse(text=paste0("lda(", target, " ~ ., data=train)")))
    test.lda <- predict(train.lda, newdata = test)
    CM <- caret::confusionMatrix(test[[target]], test.lda$class)
    sens[i] <- CM$byClass[1]
    spec[i] <- CM$byClass[2]
    accu[i] <- CM$overall[1]
    obs[1:length(folds[[i]]) + offset] <- test[[target]]
    preds[1:length(folds[[i]]) + offset] <- as.numeric(levels(test.lda$class))[test.lda$class]
    offset <- offset + length(folds[[i]])
  }
  # Plot the ROC curve
  roc(obs, preds, ci=TRUE)
  plot.roc(obs, preds, ci=TRUE, col="red")
}
# Checking covariance of each class
X <- top.data.imputed[,-1]
print(paste("Covariance of no_shock samples", cov(x = X[X$aim1_status=="no_shock"])))
print(paste("Covariance of pressor_fluid samples", cov(x = X[X$aim1_status=="pressor_fluid"])))

# Using top.data.imputed
lda_sepsis(data = top.data.imputed, target = "aim1_status", n=3)
# Using full.data, usually too heavy to run
# lda_sepsis(data = full.data, target = "aim1_status", n=3)
```

The area under the curve for classifying pressor vs. no_shock using LDA is 0.5227 only. This could be because the eset we use didn't represent the entire expression set well enough. 

With held-out data for CV, LDA gives a significantly better result as expected.
```{r eval=TRUE, echo = TRUE, results='asis'}
print("Meanwhile, held-out CV gives the result:\n")
pressor.lda <- lda(aim1_status ~ ., data=top.data.imputed, CV=TRUE)
confusion <- caret::confusionMatrix(top.data$aim1_status, pressor.lda$class)
print(paste("Sensitivity: ", confusion$byClass[1]))
print(paste("Specificity: ", confusion$byClass[2]))
print(paste("Accuracy: ", confusion$overall[1]))
```

## Random Forests results
Applying Random Forests for aim1_status classfication
3-fold CV and ROC curve for Random Forests

``` {r eval = TRUE, echo = TRUE, results = 'asis', message=FALSE, warning=FALSE}
library(randomForest)
# Cross-Validation
random_forest <- function(data, target, n){
  set.seed(252)
  N = nrow(data)
  pred_outputs.rf <- obs_outputs <- vector(mode="numeric", length=N)
  offset <- 0
  folds <- createFolds(data[[target]], k = n, list = TRUE, returnTrain = FALSE)
  for(i in 1:n){
  	train <- data[-folds[[i]],]
    test <- data[folds[[i]],]
    obs_outputs[1:length(folds[[i]]) + offset] <- test[[target]]
    # RF train/test
    rf <- eval(parse(text=paste0("randomForest(", target, "~., data=train, ntree=100)")))
  	rf.pred.curr <- predict(rf, newdata=test, type="response") 
  	pred_outputs.rf[1:length(folds[[i]]) + offset] <- rf.pred.curr
  	offset <- offset + length(folds[[i]])
  }
  print(paste("ROC plot: Random Forests for", target))
  roc(obs_outputs, pred_outputs.rf, ci=TRUE)
  plot.roc(obs_outputs, pred_outputs.rf, ci=TRUE, col="red")
}

random_forest(data = top.data.imputed, target="aim1_status", n=3)
print("Random Forests with clinical variables only")
random_forest(data = full.data.imputed[,1:17], target = "aim1_status", n = 3)
```

## KNN results
Trying out KNN with 1 nearest neighbor

``` {r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
library(FNN)
knn_run <- function(data, target, n){
  set.seed(1)
  folds <- createFolds(data[[target]], k = n, list = TRUE, returnTrain = FALSE)
  N <- nrow(data)
  preds <- obs <- vector(mode="numeric", length = N)
  offset <- 0
  for (i in 1:n){
    train <- data[-folds[[i]],]
    test <- data[folds[[i]],]
    res <- knn(train, test, train[[target]], k=1, prob=FALSE, algorithm=c("kd_tree", "cover_tree", "brute"))
    obs[1:length(folds[[i]]) + offset] <- test[[target]]
    preds[1:length(folds[[i]]) + offset] <- as.numeric(levels(res))[res]
    offset <- offset + length(folds[[i]])
  }
  # Plot the ROC curve
  print(paste("KNN plot for", target))
  roc(obs, preds, ci=TRUE)
  plot.roc(obs, preds, ci=TRUE, col="red")
}
knn_run(data = top.data.imputed, target = "aim1_status", n=3)

```


## Stratified analysis for viral and bacterial for aim1

Observing that no viral infection cases are treated with pressor, we have collapsed pressor and fluid_responsive samples together, in contrast to no_shock. Here we conduct stratified analysis of pressor_fluid vs. no_shock for patients in each group viral and bacterial 

``` {r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
# A function that takes in a expression set, and choice (a vector of 2 variables, one target, one adjusted covariant) and make heatmaps for statified analysis using interested variables in the covariant

make_heatmap <- function(gset, choice){
  label <- unique(pData(gset)$aim2_short)
  print(paste("Stratified analysis for",choice,"with", label, "samples"))
  colors <- rainbow(length(unique(gset[[choice])))
  genes <- sapply(row.names(exprs(gset)), function(x){gene_names[[x]]})
  status.colors <- unlist(lapply(as.numeric(factor(gset[[choice])), function(x){colors[x]}))
  heatmap(exprs(gset), col=viridis(256, option="B"), margins = c(5,10), ColSideColors=status.colors, labRow=genes, labCol=FALSE)
  par(lend = 1)           # square line ends for the color legend
  title(paste(choice[1], "adjusted for", choice[2], label, "only"))
  legend("topright",      # location of the legend on the heatmap plot
        legend = names(table(gset[[choice])), # category labels
        col = colors,  # color key
        lty= 1,             # line style
        lwd = 10            # line width
        )
}
# Choose the top.eset which target is pressor vs. no_shock
list.esets <- list(
  top.eset[, which(pData(top.eset)$aim2_short=="viral")], 
  top.eset[, which(pData(top.eset)$aim2_short=="bacterial")])
lapply(list.esets, make_heatmap, choice="aim1_status")

```

## Using Sparse LDA
While LDA produces low-dimensional projection of data onto most discriminative directions, in the case of data with p >> n, within-class covariance matrix of the features is singular. A method to overcome this problem is using a regularized estimate of the within-class covariance matrix in LDA (e.g. using the LASSO). Since the decision between treatment (fluid, shock, pressor, no shock) doesn't really have "clear boundaries", we applied sparse mixture discriminant analysis for more flexibility.

``` {r eval=TRUE, echo = TRUE, results='asis',message=FALSE, warning=FALSE}
library(sparseLDA)
sparse_lda <- function(data, target, lambda, n){
  folds <- createFolds(data[[target]], k=n, list=TRUE, returnTrain=FALSE)
  preds <- obs <- vector(mode="numeric", length=nrow(data))
  offset <- 0
  for (i in 1:n){
    train <- data[-folds[[i]],]
    test <- data[folds[[i]],]
    y <- model.matrix(~-1+factor(train[,1]))
    labels <- names(table(train[[target]]))
    colnames(y)[1:length(labels)] <- labels
    model.sda <- sda(as.matrix(train[,-1]), y, Z=NULL, Rj=NULL,lambda = lambda, stop=-15, maxIte=50)
    obs[1:length(folds[[i]]) + offset] <- test[[target]]
    preds[1:length(folds[[i]]) + offset] <- predict(model.sda, as.matrix(test[,-1]))$class
    offset <- offset + length(folds[[i]])
  }
  print(confusionMatrix(obs, preds))
  print(paste("Sparse Mixture LDA for", target))
  roc(obs, preds, ci=TRUE)
  plot.roc(obs, preds, col="blue")
}
                         
sparse_lda(top.data.imputed, target = "aim1_status", lambda = 3, n=3)
```

## Feature selection with lasso then apply LDA
```{r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
library(glmnet)
library(plotmo)
cvfit <- cv.glmnet(as.matrix(full.data.imputed[,-1]), full.data.imputed[,1])
#coef(cvfit, s="lambda.min")
# Top important coefficients
names(full.data.imputed)[which(abs(coef(cvfit, s="lambda.min")) >0)][-1]
sepsis.lasso <- glmnet(as.matrix(full.data.imputed[,-1]), full.data.imputed[,1], 
                       family = "binomial", lambda = cvfit$lambda.min)
print("Plotting the most important genes (highest absolute coefficient) by lasso")
par(mfrow=c(2,1))
plot_glmnet(sepsis.lasso, label=10, nresponse=1)
plot_glmnet(sepsis.lasso, label=10, nresponse=2)
```

If talking in the sense of causal inference, we can see the gene expression and the clinical variables as the confounding (L), and aim1_status as the treatment (A), we don't have knowledge about the outcome after patients receive the treatment (Y). 

Since the number of patients for each groups are quite uneven, we consider applying IPW and Matching methods. 



## Matching with nearest neighbor and mahalanobis distance 
``` {r eval=FALSE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
pressor.data <- as.data.frame(cbind(pData(noviral_eset)[,which(names(pData(noviral_eset)) %in% clin.vars)], t(exprs(noviral_eset))))
names(pressor.data)[1:length(clin.vars)] <- clin.vars
pressor.data[,1:length(clin.vars)] <- sapply(pressor.data[,1:length(clin.vars)], function(x) as.numeric(as.factor(x)))
pressor.data[,1:4] <- rfImpute(aim1_status ~., pressor.data[,1:length(clin.vars)])
pressor.data$aim1_status <- unlist(lapply(pressor.data$aim1_status, function(x){
  if (x==1){0} else {1}
}))
match1 <- matchit(aim1_status ~ sex + race_ethnicity + aim2_short,
                  data = pressor.data, method="nearest", distance="logit", reestimate=TRUE)
match2 <- matchit(aim1_status ~ sex + race_ethnicity + aim2_short,
                  data = pressor.data, method="optimal", distance="mahalanobis", reestimate=TRUE)
# Compare the results from each of the matches
summary(match1)
summary(match2)

```
# Aim1 analysis adjusted for race_ethnicity+sex and adjusted for aim2
Conduct analysis for both choices of aim1: aim1 adjusted for race_ethnicity+sex and aim1 adjusted for aim2

``` {r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
analysis(sepsis.rma, choices[[1]], groups_aim1, complete_cases = TRUE)
analysis(sepsis.rma, choices[[2]], groups_aim1, complete_cases = TRUE)
```
# Aim 2
Conduct analysis for aim2
``` {r eval=TRUE, echo = TRUE, results='asis', message=FALSE, warning=FALSE}
analysis(sepsis.rma, choices[[3]], groups_aim2, complete_cases = FALSE)
# Random Forests with full data imputed
random_forest(full.data.imputed = full.data.imputed, target = "aim2_short", n=3)
# Random Forests with clinical variables only 
clin.data <- pData(sepsis.rma)[,-which(names(pData(sepsis.rma)) %in% c("cel_file", "aim2_status",
                                                                       "aim1_combined", 
                                                                      "race_ethnicity")]
random_forest(full.data.imputed = clin.data, target = "aim2_short", n=3)

```


Knitted with:
``` {r eval=FALSE}
library(knitr)
knit2html(spin("report0915.Rmd", knit=FALSE), force_v1=TRUE)
```
